4.5 Add deprecations for most of ModelAwareTrait

markstory opened the pull request:
Adding deprecations to loadModel() is tricky. While it is important for this method to emit a deprecation so that developer can fix their usage, we also use it internally quite extensively to load ORM models.

We might need a more elaborate solution to emit deprecations for this method. One option that comes to mind is only emitting deprecations for non-ORM usage. We *could* have a `loadModel()` method in 5.0 that emits deprecations.

Refs #16974

markstory committed:
4.5 Add deprecations for most of ModelAwareTrait

Adding deprecations to loadModel() is tricky. While it is important for
this method to emit a deprecation so that developer can fix their usage,
we also use it internally quite extensively to load ORM models.

We might need a more elaborate solution to emit deprecations for this
method. One option that comes to mind is only emitting deprecations for
non-ORM usage. We *could* have a `loadModel()` method in 5.0 that emits
deprecations.

markstory committed:
Fix more tests.

ADmad commented:
I am bit conflicted about throwing deprecation errors for `loadModel()` since there isn't exactly a like for like replacement for it. For e.g. for someone using the muffin/webservice plugin the only option would be to suppress deprecation messages.

markstory commented:
> I am bit conflicted about throwing deprecation errors for loadModel() since there isn't exactly a like for like replacement for it

Same. We can only recommend using fetchTable() for ORM records. Does usemuffin have a registry that users could migrate to like we have in ElasticSearch?

ADmad commented:
Yes it does have a locator class but the combo for Crud + Webservice plugins allows one to seamlessly query API datasources just as you would do for databases by using the `ModelAwareTrait` method, simply changing the model type with `setModelType()`. Dropping the use of `ModelAwareTrait` would mean lot of additional code in Crud to switch between a `Table` instance and `Endpoint` instance of webservice.

I am kinda regretting deprecating the `ModelAwareTrait` now just to avoid the dynamic property setting it does :slightly_smiling_face:.

markstory commented:
> I am kinda regretting deprecating the ModelAwareTrait now just to avoid the dynamic property setting it does

Well we can reverse that decision :smile: If the problem was the dynamic properties perhaps that is the only part we should deprecate? We could still deprecate `loadModel()` and add a `fetchModel()` that works like `fetchTable()` 🤷

ADmad commented:
> If the problem was the dynamic properties...

Another reason for using fetchTable() instead of loadModel() was to improve type inference for static analyzers.

> We could still deprecate loadModel() and add a fetchModel() that works like fetchTable()

That's a solution I was thinking of too 🙂

markstory commented:
> That's a solution I was thinking of too 🙂

Ok. I'll revise this pull request to do that then.

markstory committed:
Add fetchModel() and un-deprecate loadModel()

When figuring out how we could add deprecations to loadModel() we
uncovered that this would be very hard because of how loadModel() is
used. The deprecation of loadModel() was primarily driven by wanting to
avoid the 8.2 dynamic property deprecation from PHP. We can add
a deprecation ourselves for that problem with a suggestion on how to
resolve this problem.

Deprecating loadModel() would also create a rough upgrade path to other
locator based implementations (like in usemuffin/webservices). Instead
we can retain loadModel() and add fetchModel(). Throughout 5.x we can
encourage users stop using loadModel(), and start using other methods.

markstory committed:
Fix deprecations

markstory committed:
Update psalm baseline.

markstory committed:
Fix logic in fetchModel.

markstory committed:
Improve tests, remove some mocks and deprecated blocks

markstory committed:
Fix phpcs

markstory committed:
Bump chronos as RC1 was missing stuff.

markstory commented:
@ADmad What do you think of this?

markstory commented:
I'll get the docs updated later this week.
